azure/functions/_http_asgi.py

```diff
@@ -0,0 +1,134 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import asyncio
+from typing import Callable, Dict, List, Tuple, Optional, Any, Union
+import logging
+from wsgiref.headers import Headers
+
+from ._abc import Context
+from ._http import HttpRequest, HttpResponse
+from ._http_wsgi import WsgiRequest
+
+
+class AsgiRequest(WsgiRequest):
+    _environ_cache: Optional[Dict[str, Any]] = None
+
+    def __init__(self, func_req: HttpRequest,
+                 func_ctx: Optional[Context] = None):
+        self.asgi_version = "2.1"
+        self.asgi_spec_version = "2.1"
+        self._headers = func_req.headers
+        super().__init__(func_req, func_ctx)
+
+    def _get_encoded_http_headers(self) -> List[Tuple[bytes, bytes]]:
+        return [(k.encode("utf8"), v.encode("utf8"))
+                for k, v in self._headers.items()]
+
+    def _get_server_address(self):
+        if self.server_name is not None:
+            return (self.server_name, int(self.server_port))
+        return None
+
+    def to_asgi_http_scope(self):
+        return {
+            "type": "http",
+            "asgi.version": self.asgi_version,
+            "asgi.spec_version": self.asgi_spec_version,
+            "http_version": "1.1",
+            "method": self.request_method,
+            "scheme": "https",
+            "path": self.path_info,
+            "raw_path": self.path_info.encode("utf-8"),
+            "query_string": self.query_string.encode("utf-8"),
+            "root_path": self.script_name,
+            "headers": self._get_encoded_http_headers(),
+            "server": self._get_server_address(),
+            "client": None
```

Comment:
Let's add a few more Azure Functions metadata fields `azure_functions.function_directory`, `azure_functions.function_name` and `azure_functions.invocation_id` for the ease of debug in the future. https://github.com/Azure/azure-functions-python-library/blob/dev/azure/functions/_http_wsgi.py#L80-L82